### PR TITLE
[modules-core][Android] Add a way to override `DevSupportManagerFactory`

### PR DESCRIPTION
**Pull request by lukmccall**
# Why

Part of https://linear.app/expo/issue/ENG-4210/use-reactnativehostwrapper-to-replace-devsupportmanagerfactory.
Adds a way to override `DevSupportManagerFactory` in the module. This change was a little bit tricky, cause the method to override was introduced in React Native 0.67 and we still want to have backwards compatibility with previous RN versions. 

# How

- Added a `reactNativeHostWrapper` and `legacyReactNativeHostWrapper` directories. Ony one of them is added to the sources - the choice depends on the detected RN version. 
- Added `getDevSupportManagerFactory` to the `ReactNativeHostHandler` interface. Note that I didn't specify the return type here. However, we can do it, but I don't think that is worth it.  We would have to have two different versions of this interface and choose between them by the RN version.
  
# Test Plan

- compile bare-expo with RN 66 and 67 ✅

### CODE REVIEW

**Comment by Kudo** [comment]
i am getting little afraid `legacyReactNativeHostWrapper` and `reactNativeHostWrapper` out of sync. how do you think if we use reflection to check the `getDevSupportManagerFactory()` method exist in new react-native? i think we already had some reflections there 😅

**Comment by lukmccall** [comment]
> i am getting little afraid `legacyReactNativeHostWrapper` and `reactNativeHostWrapper` out of sync. how do you think if we use reflection to check the `getDevSupportManagerFactory()` method exist in new react-native? i think we already had some reflections there 😅

Yes, you can check if the method exists but you can't override it conditionally. So the reflection here doesn't give us anything.

**Comment by Kudo** [comment]
> Yes, you can check if the method exists but you can't override it conditionally. So the reflection here doesn't give us anything.

oh no, that's true. sorry, i still hope if we could have some way to reduce the syncing maintenance cost. how do you think if we have a common base class?

```kotlin
open class ReactNativeHostWrapperBase(
  application: Application,
  protected val host: ReactNativeHost
) : ReactNativeHost(application) {
  // original ReactNativeHostWrapper code
}

// packages/expo/android/src/legacyReactNativeHostWrapper/expo/modules/ReactNativeHostWrapper.kt
class ReactNativeHostWrapper(
  application: Application,
  host: ReactNativeHost
) : ReactNativeHostWrapperBase (application, host) {
}

// packages/expo/android/src/reactNativeHostWrapper/expo/modules/ReactNativeHostWrapper.kt
class ReactNativeHostWrapper(
  application: Application,
  host: ReactNativeHost
) : ReactNativeHostWrapperBase (application, host) {
  override fun getDevSupportManagerFactory(): DevSupportManagerFactory? {
    return null
  }
}
```

File: packages/expo/android/build.gradle [hunk 1]
```diff
@@ -1,10 +1,41 @@
...
+      "-e",
+      "console.log(require.resolve('react-native/package.json'));"
+  ]
+      .execute([], projectDir)
+      .text
+      .trim()
+
+  def packageJsonFile = new File(reactNativePackageJsonPath)
+  def packageJson = new JsonSlurper().parseText(packageJsonFile.text)
+  def version = safeExtGet("reactNativeVersion", packageJson.version)
```
Comment (Kudo) [inline packages/expo/android/build.gradle:26]: neat: i found @esamelson's approach to [get the version directly from nodejs](https://github.com/expo/expo/pull/16441/files#diff-4fa7841d1fd49df86f44c26076d9dc6752f3c183c2b659e1243b8b8c3d223f41R157-R159) is clever. maybe it's worth to take a look.

**Review by Kudo — APPROVED** [review]
thanks for this enhancement 👏

**Review by esamelson — APPROVED** [review]
👌

**Pull request merged**